bees/facebookbee/facebookbee.go

```diff
@@ -71,30 +80,212 @@ func (mod *FacebookBee) Run(eventChan chan bees.Event) {
+	baseURL := "https://graph.facebook.com/oauth/access_token"
+	v := url.Values{}
+	v.Set("grant_type", "fb_exchange_token")
+	v.Set("client_id", mod.clientID)
+	v.Set("client_secret", mod.clientSecret)
+	v.Set("fb_exchange_token", mod.accessToken)
+	graphUrl := baseURL + "?" + v.Encode()
+
+	res, err := http.Get(graphUrl)
+
+	if err != nil || res == nil {
+		mod.LogErrorf("Fetching long lived user access token failed: %v", err)
+		return ""
+	}
+
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
+
+	if err != nil {
+		mod.LogErrorf("Reading content while fetching long lived user access token failed: %v", err)
+		return ""
+	}
+
+	// mod.Logf("Long lived user access token result: \"%s\"", body)
+
+	type RequestResult struct {
+		AccessToken string `json:"access_token"`
+	}
+
+	var tokenRes RequestResult
+	err = json.Unmarshal(body, &tokenRes)
+
+	if err != nil {
+		mod.LogErrorf("Parsing result while fetching long lived user access token failed: %v", err)
+		return ""
+	}
+
+	return tokenRes.AccessToken
+}
+
+func (mod *FacebookBee) fetchAccountID(accessToken string) string {
+	baseURL := "https://graph.facebook.com/v8.0/me"
+	v := url.Values{}
+	v.Set("access_token", accessToken)
+	v.Set("fields", "id")
+	graphUrl := baseURL + "?" + v.Encode()
+
+	res, err := http.Get(graphUrl)
+
+	if err != nil || res == nil {
+		mod.LogErrorf("Fetching user id failed: %v", err)
+		return ""
+	}
+
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
+
+	if err != nil {
+		mod.LogErrorf("Reading content while fetching user id failed: %v", err)
+		return ""
+	}
+
+	// mod.Logf("user id result: \"%s\"", body)
+
+	type RequestResult struct {
+		ID string `json:"id"`
+	}
+
+	var tokenRes RequestResult
+	err = json.Unmarshal(body, &tokenRes)
+
+	if err != nil {
+		mod.LogErrorf("Parsing result while fetching user id failed: %v", err)
+		return ""
+	}
+
+	return tokenRes.ID
+}
+
+func (mod *FacebookBee) fetchPermanentPageAccessToken(accountID string, accessToken string) string {
+	// the method in https://developers.facebook.com/docs/pages/access-tokens/#get-a-page-access-token doesn't work!
+	// https://github.com/Bnjis/Facebook-permanent-token-generator/blob/master/src/components/Form.js helped a lot
+	baseURL := "https://graph.facebook.com/v8.0/" + accountID + "/accounts"
+	v := url.Values{}
+	v.Set("access_token", accessToken)
+	graphUrl := baseURL + "?" + v.Encode()
+
+	// var buf io.ReadWriter
+	res, err := http.Get(graphUrl)
+
+	if err != nil || res == nil {
+		mod.LogErrorf("Fetching page token failed: %v", err)
+		return ""
+	}
+
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
+
+	if err != nil {
+		mod.LogErrorf("Reading content while fetching page token failed: %v", err)
+		return ""
+	}
+
+	// mod.Logf("Page token result: \"%s\"", body)
+
+	type RequestResultInner struct {
+		AccessToken string `json:"access_token"`
+	}
+
+	type RequestResult struct {
```

Comment:
You could declare the two structs inline:

```
type RequestResult struct {
    Data []struct {
        AccessToken string `json:"access_token"
    } `json:"data"`
}
```

Reply:
Great idea, thank you!

Reply:
a ` was missing 😉